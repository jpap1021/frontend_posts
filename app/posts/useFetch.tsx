import React, { useReducer, useEffect, useState } from "react";

interface Post {
  id: number;
  title: string;
  body: string;
  comments?: string[];
}

export interface State {
  data:
    | (Post | { comments: any[]; id: number; title: string; body: string })[]
    | undefined;
  isLoading: boolean;
  error: boolean;
}

export interface Action {
  type: string;
  payload?: any;
  comment?: string;
}

export const useFetch = ({
  dispatch,
}: any): {
  handleAddComment: (postId: number, comment: string) => void;
} => {
  const [initialStateLoaded, setInitialStateLoaded] = useState(false);

  const handleAddComment = (postId: number, comment: string) => {
    dispatch({ type: "ADD_COMMENT", payload: { postId, comment } });
  };

  useEffect(() => {
    const fetchData = async () => {
      dispatch({ type: "FETCH_START" });

      try {
        const response = await fetch(
          "https://jsonplaceholder.typicode.com/posts"
        );
        const jsonData = await response.json();
        dispatch({ type: "FETCH_SUCCESS", payload: jsonData });
      } catch (error) {
        dispatch({ type: "FETCH_ERROR" });
      }
    };

    if (!initialStateLoaded) {
      fetchData();
      setInitialStateLoaded(true);
    }
  }, [initialStateLoaded, dispatch]);

  return { handleAddComment };
};
