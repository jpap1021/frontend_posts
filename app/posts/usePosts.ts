"use client";
import { useRouter } from "next/navigation";
import { useReducer } from "react";
import { reducer } from "../reducer/reducer";
import { State, useFetch } from "./useFetch";

const initialState: State = {
  data: undefined,
  isLoading: false,
  error: false,
};

export const usePost = () => {
  const [state, dispatch] = useReducer(reducer, initialState);

  const { handleAddComment } = useFetch({ state, dispatch });
  const router = useRouter();

  const posts = state?.data;
  const addComment = (id: number) => {
    router.push(`comments/${id}`, {});
  };
  return {
    posts,
    addComment,
    handleAddComment,
  };
};
