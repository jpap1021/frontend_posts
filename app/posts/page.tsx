"use client";
import Box from "@mui/material/Box";
import Button from "@mui/material/Button";
import AddIcon from "@mui/icons-material/Add";
import { usePost } from "./usePosts";

export default function PostPage() {
  const { posts, addComment } = usePost();

  return (
    <div>
      <Box
        sx={{
          display: "flex",
          justifyContent: "center",
          flexDirection: "column",
          alignContent: "center",
          width: "100%",
          boxShadow: "",
        }}
      >
        {posts?.map((post: any) => (
          <Box
            sx={{
              width: 500,
              height: "auto",
              boxShadow: "0 2px 4px rgba(0, 0, 0, 0.2)",
              margin: "20px auto",
              padding: "20px",
            }}
            key={post.id}
          >
            <Box
              sx={{
                fontWeight: "bold",
                fontSize: "18px",
                marginBottom: "10px",
              }}
            >
              {post.title}
            </Box>
            <Box>{post.body}</Box>
            <Box
              sx={{
                display: "flex",
                justifyContent: "flex-end",
                marginTop: "20px",
              }}
            >
              <Button
                variant="contained"
                startIcon={<AddIcon />}
                onClick={() => addComment(post.id)}
              >
                Add Comment
              </Button>
            </Box>
          </Box>
        ))}
      </Box>
    </div>
  );
}
