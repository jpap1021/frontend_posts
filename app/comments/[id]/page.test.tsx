import { render, screen, fireEvent } from "@testing-library/react";
import CommentPage from "@/app/comments/[id]/page";
import { usePost } from "../../posts/usePosts";

jest.mock("../../posts/usePosts");

describe("CommentPage", () => {
  it("snapshot", () => {
    const postId = 1;
    const comments = ["Comment 1", "Comment 2"];
    (usePost as jest.Mock).mockImplementation(() => ({
      posts: [{ id: postId, comments }],
      handleAddComment: jest.fn(),
    }));
    const { container } = render(<CommentPage params={{ id: postId }} />);
    expect(container).toMatchSnapshot();
  });

  it("updates inputComment state when the textarea value changes", () => {
    const postId = 1;
    const handleAddCommentMock = jest.fn();
    (usePost as jest.Mock).mockImplementation(() => ({
      posts: [],
      handleAddComment: handleAddCommentMock,
    }));

    render(<CommentPage params={{ id: postId }} />);
    const openButton = screen.getByRole("button", { name: "Add Comment" });
    fireEvent.click(openButton);

    const textarea = screen.getByTestId("comment-textarea");
    const form = screen.getByTestId("comment-form");

    fireEvent.change(textarea, { target: { value: "This is a comment" } });

    fireEvent.submit(form);

    expect(handleAddCommentMock).toHaveBeenCalledWith(
      postId,
      "This is a comment"
    );
  });
});
