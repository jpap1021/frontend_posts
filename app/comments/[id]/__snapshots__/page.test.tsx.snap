// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentPage snapshot 1`] = `
<div>
  <div
    class="container_comments"
  >
    <h1>
      Comments Post id # 
      1
    </h1>
    <div
      class="comments"
    >
      Comment 1
    </div>
    <div
      class="comments"
    >
      Comment 2
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1465nk4-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Add Comment
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
